### ISSUE TITLE
머딥 - 13d (선형 분류, 트리 이론)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#43]
# 목차
1) 선형 분류
>1. 로지스틱 회귀
>2. SVM
>3. Softmax

2) 트리
>4. 랜덤 포레스트
>5. 그라디언트 부스팅

3) 기타
> 6. 팁

## 선형 분류
feature 개수가 많을 수록 잘 분류한다
feature 개수가 적을 수록 못 분류한다

**C : 오차 규제 (오차를 허용하는 않는 정도)**
> C > 0, 상수
> 높아질 수록 데이터 하나하나에 집중한다
> 높아질 수록 아웃라이어(**초평면 안 데이터**, 잘못 분류된 데이터)를 포함한다
> 높아질 수록 가중치 **w**가 올라간다
> 높아질 수록 복잡도가 올라간다

### 1. 로지스틱 회귀
하이퍼 파라미터 : max_iter
> 최대 학습 횟수 = 미분 횟수

### 2. 서포트 벡터 머신
linear 모드와 rbf(가우시안 커널) 모드가 있다
> **커널**? 마진 폭
> gamma? 폭의 너비를 조절하는 하이퍼파라미터 (반비례 관계)

서포트 벡터 간의 거리(margin)을 최대한 높이도록  hyperplane을 그린다
> 서포트 벡터 = hyperplane에 붙은 y_train

손실함수에서 크사이를 줄여야 마진이 커짐

SVR에서 C는 penalty의 alpha를 의미하는 게 아니다. 그저 선형분류의 C를 의미한다.

초평면 안에 있는 데이터는 오차이자 아웃라이어다. 잘못 분류된 데이터 또한 마찬가지이다.

### 3. SoftMax
분자와 분모에 있는 함수 해석
> W_class * X_row + bias
> W_k 는 행벡터 형태다

자연상수를 쓰는 이유? **비선형**으로 사용하려고 (**복잡도 up**)

가중치 행렬해석
|   | X1 feature | X2 feature |
|--|-------------|------------|
| 0 | 0번클래스 | ~0번클래스 |
| 1 | 1번클래스 | ~1번클래스 |
| 2 | 2번클래스 | ~2번클래스 |
| --| 양성 | 음성 |
> OVA 방식이다

## 트리 모델
복잡도 조절 방식
[1] pre_pruning = max_depth
[2] post_pruning = 데이터 포인트가 적은 노드 삭제

Feature importance
:부모노드에서 pure 노드까지 가는데  걸리는 가중치 w
> 이것만으로 feature의 중요도를 파악하는 것은 섣부른 판단이다
> b/c 질문지를 만들 때 모든 feature를 사용하는게 아니기 때문

시계열 분석
트리모델은 **불가**

### 4.  랜덤 포레스트
overfitting tree
무작위 특성(bootstrap)으로 추출해서 트리마다 각각 예측한다
n_esitimators : 트리 개수
max_features : 나무에 사용할 최대 가지 수 (자동)
> max_feature = 1이면? 하나의 feature만 중복해서 뽑는다

약점 : text data에 약하다

### 5. 그라디언트 부스팅
underfitting tree
둘 다 선택되는 열이 있고, 아닌 열이 있다
n_estimators : 트리 개수
max_depth : 나무 깊이 (default=3) 1~5개

서로 관련 있는 feature가 많을 때 효과적이다.
n_estimator가 많으면 overfitting 일어난다
